File: test/test_remove.cpp
#include <stdio.h>
#include "TRTCContext.h"
#include "DVVector.h"
#include "remove.h"

int main()
{
	
	{
		int h_value[6] = { 3, 1, 4, 1, 5, 9 };
		DVVector d_value("int32_t", 6, h_value);
		uint32_t count = TRTC_Remove(d_value, DVInt32(1));
		d_value.to_host(h_value, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_value[i]);
		puts("");
	}

	{
		int h_in[6] = { -2, 0, -1, 0, 1, 2 };
		DVVector d_in("int32_t", 6, h_in);
		int h_out[6];
		DVVector d_out("int32_t", 6);
		uint32_t count = TRTC_Remove_Copy(d_in, d_out, DVInt32(0));
		d_out.to_host(h_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_out[i]);
		puts("");
	}

	Functor is_even = { {},{ "x" }, "        return x % 2 == 0;\n" };

	{
		int h_value[6] = { 1, 4, 2, 8, 5, 7 };
		DVVector d_value("int32_t", 6, h_value);
		uint32_t count = TRTC_Remove_If(d_value, is_even);
		d_value.to_host(h_value, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_value[i]);
		puts("");
	}

	{
		int h_in[6] = { -2, 0, -1, 0, 1, 2 };
		DVVector d_in("int32_t", 6, h_in);
		int h_out[6];
		DVVector d_out("int32_t", 6);
		uint32_t count = TRTC_Remove_Copy_If(d_in, d_out, is_even);
		d_out.to_host(h_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_out[i]);
		puts("");
	}

	Functor identity("Identity");

	{
		int h_value[6] = { 1, 4, 2, 8, 5, 7 };
		DVVector d_value("int32_t", 6, h_value);
		int h_stencil[6] = { 0, 1, 1, 1, 0, 0 };
		DVVector d_stencil("int32_t", 6, h_stencil);
		uint32_t count = TRTC_Remove_If_Stencil(d_value, d_stencil, identity);
		d_value.to_host(h_value, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_value[i]);
		puts("");
	}

	{
		int h_in[6] = { -2, 0, -1, 0, 1, 2 };
		DVVector d_in("int32_t", 6, h_in);
		int h_stencil[6] = { 1, 1, 0, 1, 0, 1 };
		DVVector d_stencil("int32_t", 6, h_stencil);
		int h_out[6];
		DVVector d_out("int32_t", 6);
		uint32_t count = TRTC_Remove_Copy_If_Stencil(d_in, d_stencil, d_out, identity);
		d_out.to_host(h_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_out[i]);
		puts("");
	}

	return 0;
}

File: CSharp/test/test_remove/test_remove.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ThrustRTCSharp;

namespace test_remove
{
    class test_remove
    {
        static void Main(string[] args)
        {
            {
                DVVector d_value = new DVVector(new int[] { 3, 1, 4, 1, 5, 9 });
                long count = TRTC.Remove(d_value, new DVInt32(1));
                print_array((int[])d_value.to_host(0, count));
            }

            {
                DVVector d_in = new DVVector(new int[] { -2, 0, -1, 0, 1, 2 });
                DVVector d_out = new DVVector("int32_t", 6);
                long count = TRTC.Remove_Copy(d_in, d_out, new DVInt32(0));
                print_array((int[])d_out.to_host(0, count));
            }

            Functor is_even = new Functor(new string[] { "x" }, "        return x % 2 == 0;\n");

            {
                DVVector d_value = new DVVector(new int[] { 1, 4, 2, 8, 5, 7 });
                long count = TRTC.Remove_If(d_value, is_even);
                print_array((int[])d_value.to_host(0, count));
            }

            {
                DVVector d_in = new DVVector(new int[] { -2, 0, -1, 0, 1, 2 });
                DVVector d_out = new DVVector("int32_t", 6);
                long count = TRTC.Remove_Copy_If(d_in, d_out, is_even);
                print_array((int[])d_out.to_host(0, count));
            }

            Functor identity = new Functor("Identity");

            {
                DVVector d_value = new DVVector(new int[] { 1, 4, 2, 8, 5, 7 });
                DVVector d_stencil = new DVVector(new int[] { 0, 1, 1, 1, 0, 0 });
                long count = TRTC.Remove_If_Stencil(d_value, d_stencil, identity);
                print_array((int[])d_value.to_host(0, count));
            }

            {
                DVVector d_in = new DVVector(new int[] { -2, 0, -1, 0, 1, 2 });
                DVVector d_stencil = new DVVector(new int[] { 1, 1, 0, 1, 0, 1 });
                DVVector d_out = new DVVector("int32_t", 6);
                long count = TRTC.Remove_Copy_If_Stencil(d_in, d_stencil, d_out, identity);
                print_array((int[])d_out.to_host(0, count));
            }
        }

        static void print_array<T>(T[] arr)
        {
            foreach (var item in arr)
            {
                Console.Write(item.ToString());
                Console.Write(" ");
            }
            Console.WriteLine("");
        }
    }
}
